Lib/unittest/mock.py

```diff
@@ -2482,6 +2484,55 @@ def __set__(self, obj, val):
         self(val)
 
 
+class WaitableMock(MagicMock):
+    """
+    A mock that can be used to wait until it was called.
+
+    `event_class` - Class to be used to create event object.
+    Defaults to Threading.Event and can take values like multiprocessing.Event.
+    """
+
+    def __init__(self, *args, event_class=threading.Event, **kwargs):
+        _safe_super(WaitableMock, self).__init__(*args, **kwargs)
+        self._event = event_class()
+        self._expected_calls = {}
+
+    def _mock_call(self, *args, **kwargs):
+        ret_value = _safe_super(WaitableMock, self)._mock_call(*args, **kwargs)
+
+        for call in self._mock_mock_calls:
+            event = self._expected_calls.get(call.args)
+            if event and not event.is_set():
+                event.set()
+
+        self._event.set()
+
+        return ret_value
+
+    def wait_until_called(self, timeout=1.0):
+        """Wait until the mock object is called.
+
+        `timeout` - time to wait for in seconds. Defaults to 1.
+        """
+        return self._event.wait(timeout=timeout)
+
+    def wait_until_called_with(self, *args, timeout=1.0):
+        """Wait until the mock object is called with given args.
+        If args is empty then it waits for the mock object to be called.
```

Comment:
I would allow for arts to be empty, which means called with no args. Otherwise users have no way to wait for that.

If they want to wait for any call they can use the other method.

Reply:
Makes sense. I have removed this line and code to default to `self._event` during empty args. Previously, calling the function without args would create an empty tuple as key in `self._expected_calls` and set the event along with `self._event`. So it was set before too but now as per suggestion instead of using `self._event` the appropriate event object at `self._expected_calls[()]` is used .